### PR TITLE
Add selector concurrency property tests (#57)

### LINKED ISSUE #57: Add property-based and model-checking concurrency tests for _EnvBackendSelector
## Background

PR #55 introduces three concurrency invariants in `benchmarks/tee_profile_worker.py` covering reentrancy, cross-thread environment isolation, and lock serialisation. The current test suite exercises only fixed-point examples; it does not cover arbitrary thread counts, backend name sequences, or execution orderings.

Requested by @leynos in https://github.com/leynos/cuprum/pull/55.

## Desired changes

- Add [Hypothesis](https://hypothesis.readthedocs.io/)-driven property tests for `_EnvBackendSelector` reentrancy, parameterised over backend names and execution orderings.
- Parameterise existing concurrency tests over thread count (e.g. 2–8 threads) and backend combinations.
- Investigate model-checking approaches (e.g. `threading`-level interleaving analysis) for the lock-acquisition and environment-mutation state machine.

## Acceptance criteria

- At least one Hypothesis test that generates arbitrary sequences of backend names and asserts reentrancy raises `RuntimeError` on nested same-thread entry.
- At least one Hypothesis test that generates arbitrary thread counts and asserts all workers complete with `status == "ok"` and `exit_code == 0`.
- No regressions in existing tests.

### PR DESCRIPTION
**Pull request by leynos**
## Summary

This branch adds property-based and checkpointed concurrency coverage for issue #57. It generates backend selector reentrancy cases, exercises concurrent worker completion across generated backend sequences and thread counts, and documents the `_EnvBackendSelector` state-machine invariants that the tests exercise.

Closes #57.

## Review walkthrough

- Start with [cuprum/unittests/test_tee_profile_worker.py](https://github.com/leynos/cuprum/blob/issue-57-add-property-based-and-model-checking-concurrency-tests-for-envbackendselector/cuprum/unittests/test_tee_profile_worker.py) to see the Hypothesis strategies, generated selector tests, and checkpointed interleaving test.
- Review the module docstring in [cuprum/unittests/test_tee_profile_worker.py](https://github.com/leynos/cuprum/blob/issue-57-add-property-based-and-model-checking-concurrency-tests-for-envbackendselector/cuprum/unittests/test_tee_profile_worker.py) for the documented selector invariants and model-checking notes.

## Validation

- `uv run pytest cuprum/unittests/test_tee_profile_worker.py`: 20 passed.
- `make check-fmt`: passed.
- `make lint`: passed.
- `make typecheck`: passed.
- `make test`: 474 passed, 43 skipped.

## Notes

- `coderabbit review --agent` was attempted twice after the implementation milestone. The CLI returned a recoverable service-side rate limit both times, so there were no CodeRabbit concerns to clear locally.

## Summary by Sourcery

Add property-based and checkpointed concurrency coverage for the _EnvBackendSelector used by tee_profile_worker, and reorganize its test suite into focused core, CLI, and concurrency modules.

Enhancements:
- Document _EnvBackendSelector concurrency invariants and model-checking considerations in the concurrency test module.
- Refactor tee_profile_worker tests into separate core execution, CLI/config validation, and concurrency-focused files for clearer coverage boundaries.

Tests:
- Introduce Hypothesis-based generation of nested backend selector pairs and concurrent backend sequences to validate selector reentrancy and race-freedom across backends and thread counts.
- Add checkpointed interleaving tests that assert environment mutation and observation are correctly serialized by the backend selector lock.
- Preserve and relocate existing worker core-behavior, CLI, and configuration-validation tests into dedicated modules, updating associated snapshots.

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai please create new GitHub issues for the following:

**High-Value Candidates**

CrossHair is only a good fit where the target can be made pure and bounded. For async subprocess, file descriptor, or OS-environment code, Hypothesis with fakes is the better immediate tool; CrossHair becomes useful after extracting pure state-transition helpers.

| Area | Tool | Justification | Benefit | Cost Trade-Off | Verifiability Improvements |
|---|---:|---|---|---|---|
| [_emit_completed_lines / _strip_line_ending](/home/leynos/.lody/repos/github---leynos---cuprum/worktrees/614e3091-8d73-404f-8c2c-a5022f5c0087/cuprum/_streams.py:202) | Hypothesis + CrossHair | Pure line-splitting logic currently covered mostly through stream behaviour. | Proves no dropped text, stable newline handling, correct final partial-line behaviour. | Low. Fast, pure, good shrinking. | Extract `_split_complete_lines(text) -> (lines, remainder)` with explicit contract. |
| [CuprumContext.narrow](/home/leynos/.lody/repos/github---leynos---cuprum/worktrees/614e3091-8d73-404f-8c2c-a5022f5c0087/cuprum/context.py:159) and [_validate_timeout](/home/leynos/.lody/repos/github---leynos---cuprum/worktrees/614e3091-8d73-404f-8c2c-a5022f5c0087/cuprum/context.py:42) | Hypothesis, CrossHair for pure helpers | Allowlist intersection, hook ordering, and timeout inheritance have many small combinations. | Prevents subtle permission broadening and hook-order regressions. | Low. Mostly dataclass/pure logic. | Split allowlist merge, hook merge, and timeout normalisation into named helpers. |
| [_build_final_results](/home/leynos/.lody/repos/github---leynos---cuprum/worktrees/614e3091-8d73-404f-8c2c-a5022f5c0087/cuprum/concurrent.py:166) | CrossHair + Hypothesis | Fail-fast compaction maps partial results to new failure indices. | Proves failure indices are valid, sorted, and point to failed results after cancellation. | Very low. Pure reducer. | Keep it pure; add simple postconditions/docstring invariants. |
| [ProgramCatalogue indexing](/home/leynos/.lody/repos/github---leynos---cuprum/worktrees/614e3091-8d73-404f-8c2c-a5022f5c0087/cuprum/catalogue.py:97) | Hypothesis | Duplicate project/program ownership cases are example-based. | Exercises random ownership graphs and lookup/is_allowed consistency. | Low. Pure data construction. | Return structured duplicate diagnostics from private validators. |
| [_coerce_argv / make](/home/leynos/.lody/repos/github---leynos---cuprum/worktrees/614e3091-8d73-404f-8c2c-a5022f5c0087/cuprum/sh.py:79) and [Pipeline.concat](/home/leynos/.lody/repos/github---leynos---cuprum/worktrees/614e3091-8d73-404f-8c2c-a5022f5c0087/cuprum/sh.py:424) | Hypothesis | CLI argv construction and pipeline composition are algebraic and example-heavy today. | Proves ordering, keyword normalisation, `None` rejection, associativity, and no stage loss. | Low to moderate. Need generated `SafeCmd`s. | Add pure `build_argv` helper and lightweight command factories for tests. |
| [safe_path / git_ref](/home/leynos/.lody/repos/github---leynos---cuprum/worktrees/614e3091-8d73-404f-8c2c-a5022f5c0087/cuprum/builders/args.py:52) | Hypothesis | String validators are classic fuzz targets; current tests cover curated samples. | Finds edge cases around empty strings, whitespace, NULs, traversal, suffixes, and Unicode. | Low. Some oracle design needed. | Factor validators into “classify reason” helpers so properties can assert rejection category. |
| [tar_create / rsync_sync](/home/leynos/.lody/repos/github---leynos---cuprum/worktrees/614e3091-8d73-404f-8c2c-a5022f5c0087/cuprum/builders/tar.py:51) | Hypothesis | Command construction has flag-combination and ordering risk. | Proves exactly-one compression flag, required argv shape, deterministic path conversion. | Low. Pure command vectors. | Extract immutable argv-builder helpers before wrapping in `sh.make`. |
| [_read_backend_env / get_stream_backend](/home/leynos/.lody/repos/github---leynos---cuprum/worktrees/614e3091-8d73-404f-8c2c-a5022f5c0087/cuprum/_backend.py:46) | Hypothesis | Env parsing and cached backend resolution combine normalisation, availability, and failure paths. | Proves `AUTO` never leaks from resolver, invalid values fail, forced Rust availability is respected. | Medium due cache/global env resets. | Split pure `resolve_backend(requested, rust_available)` from cached/env wrapper. |
| [_PipelineWaitState / _process_completed_task](/home/leynos/.lody/repos/github---leynos---cuprum/worktrees/614e3091-8d73-404f-8c2c-a5022f5c0087/cuprum/_pipeline_wait.py:29) | Hypothesis state machine; CrossHair after refactor | Pipeline completion ordering is a bounded state machine hidden inside async process waits. | Proves first-failure semantics, timing slot population, and termination decisions. | High. Needs fake processes/tasks. | Move completion handling to a pure transition method taking `(state, completed_idx, exit_code)`. |
| [_pipeline_streams dispatch/fd lifecycle](/home/leynos/.lody/repos/github---leynos---cuprum/worktrees/614e3091-8d73-404f-8c2c-a5022f5c0087/cuprum/_pipeline_streams.py:136) | Hypothesis with fault injection | Rust dispatch, FD blocking restore, and pipe-error suppression have many partial-failure paths. | Catches leaked blocking state, missing resume calls, wrong fallback, and swallowed unexpected errors. | Medium-high. Mock-heavy async/FD work. | Introduce small FD-state object and context manager for pause/blocking lifecycle. |
| [_handle_subprocess_timeout](/home/leynos/.lody/repos/github---leynos---cuprum/worktrees/614e3091-8d73-404f-8c2c-a5022f5c0087/cuprum/_subprocess_execution.py:263) and [_terminate_pipeline_remaining_stages](/home/leynos/.lody/repos/github---leynos---cuprum/worktrees/614e3091-8d73-404f-8c2c-a5022f5c0087/cuprum/_process_lifecycle.py:235) | Hypothesis; CrossHair on extracted transition data | Timeout and cleanup paths are temporal and branch-heavy. | Proves timeout payload consistency, idempotent cleanup, and no double-termination scheduling. | Medium-high. Needs fakes around process state. | Represent process lifecycle as explicit enum/state records, then verify reducers. |
| [benchmark comparison / ratchet reducers](/home/leynos/.lody/repos/github---leynos---cuprum/worktrees/614e3091-8d73-404f-8c2c-a5022f5c0087/benchmarks/comparison_analysis.py:229) | Hypothesis | JSON plan/result matching has ordering, duplicate, missing-pair, and numeric edge cases. | Proves grouping, sorted output, duplicate rejection, matching Rust scenario sets, and regression ratios. | Low to medium. Requires compact payload strategies. | Build typed scenario/result factories; keep validation separate from report rendering. |
| [folded stack summarisation](/home/leynos/.lody/repos/github---leynos---cuprum/worktrees/614e3091-8d73-404f-8c2c-a5022f5c0087/benchmarks/summarize_folded.py:20) | Hypothesis + CrossHair | Parser and counter aggregation already have property tests, but ranking/duplicate-frame invariants can go deeper. | Proves inclusive counts dedupe repeated frames, leaf counts sum correctly, ranking is deterministic. | Low. Mostly pure. | Extract parse/add/rank contracts as standalone helpers with explicit invariants. |
| [metrics/tracing/logging hooks](/home/leynos/.lody/repos/github---leynos---cuprum/worktrees/614e3091-8d73-404f-8c2c-a5022f5c0087/cuprum/adapters/metrics_adapter.py:206) | Hypothesis stateful | Event streams can arrive in varied phase/order/pid combinations. | Proves counters/spans/log records are created exactly when intended and active maps drain. | Medium. Needs synthetic event generators. | Extract event-to-operation reducers from side-effecting collector/span calls. |

**Recommended Order**

1. Start with pure, low-cost properties: `_emit_completed_lines`, `_build_final_results`, `CuprumContext.narrow`, catalogue indexing, and argv/pipeline composition.
2. Next cover validators/builders and benchmark reducers.
3. Treat pipeline wait, FD lifecycle, subprocess timeout, and adapters as state-machine work after small refactors isolate pure transitions.

**Comment by leynos** [comment]
@coderabbitai please create new GitHub issues for the following:

**Ranked Candidates**

Baseline: I found no Rust-level `#[test]`, `proptest`, `kani`, or `verus` coverage under `rust/cuprum-rust`. Current coverage is Python integration/scenario coverage around the extension.

| Target | Tool | Justification | Benefit | Cost | Verifiability Improvement |
|---|---:|---|---|---|---|
| [decode_utf8_replace](/home/leynos/.lody/repos/github---leynos---cuprum/worktrees/614e3091-8d73-404f-8c2c-a5022f5c0087/rust/cuprum-rust/src/utf8.rs:37) | proptest, some Kani | Correctness depends on arbitrary byte streams and chunk boundaries. Fixed tests cover only a few invalid/incomplete UTF-8 examples. | Proves Rust decoding matches `String::from_utf8_lossy` across arbitrary payloads and partitions. | Low-medium. Pure logic, good shrinkability. | Add `decode_chunks(chunks) -> (String, pending)` or a pure `decode_step` helper. |
| [handle_utf8_error](/home/leynos/.lody/repos/github---leynos---cuprum/worktrees/614e3091-8d73-404f-8c2c-a5022f5c0087/rust/cuprum-rust/src/utf8.rs:78) / [handle_incomplete_sequence](/home/leynos/.lody/repos/github---leynos---cuprum/worktrees/614e3091-8d73-404f-8c2c-a5022f5c0087/rust/cuprum-rust/src/utf8.rs:109) | Kani | Indexing and `drain(..valid_up_to + error_len)` are safe only under `Utf8Error`-derived bounds. | Exhaustively checks bounded malformed byte layouts for no panic and correct pending/output transitions. | Medium. Needs small harness or transition extraction. | Replace implicit mutation with an enum action like `ReplaceAndDrain(n)` / `KeepTail(n)`. |
| [append_valid_prefix](/home/leynos/.lody/repos/github---leynos---cuprum/worktrees/614e3091-8d73-404f-8c2c-a5022f5c0087/rust/cuprum-rust/src/utf8.rs:65) | Kani | Contains `unsafe from_utf8_unchecked`; invariant is external to the type system. | Confirms checked and unchecked prefix decoding are equivalent for bounded valid prefixes. | Low-medium. | Keep unsafe block tiny; add debug assertion or a checked helper for verification builds. |
| [validate_buffer_size](/home/leynos/.lody/repos/github---leynos---cuprum/worktrees/614e3091-8d73-404f-8c2c-a5022f5c0087/rust/cuprum-rust/src/lib.rs:119) | proptest | Current tests hit `0` and normal values, not platform boundary values. | Proves all non-positive values reject and accepted values fit `usize`. | Low. | Extract `checked_buffer_size(i64) -> Result<usize, ErrorKind>` and add an explicit upper memory cap. |
| [convert_fd](/home/leynos/.lody/repos/github---leynos---cuprum/worktrees/614e3091-8d73-404f-8c2c-a5022f5c0087/rust/cuprum-rust/src/lib.rs:90) | proptest | FD/handle conversion is platform-gated; Windows branch is not exercised here. | Catches negative, truncation, and range boundary errors before I/O starts. | Low, except CI needs platform coverage. | Use shared pure range-check helper before platform conversion. Explicitly reject negative Windows handles. |
| [pump_stream_files_readwrite](/home/leynos/.lody/repos/github---leynos---cuprum/worktrees/614e3091-8d73-404f-8c2c-a5022f5c0087/rust/cuprum-rust/src/lib.rs:209) | proptest, Kani after refactor | Writer-open state latches after non-fatal write errors; fixed tests cover only broad cases. | Proves total bytes are monotonic, writes stop after broken pipe, reader drain continues, fatal errors propagate. | Medium-high without seams. | Extract a pure pump state machine over `ReadEvent` and `WriteEvent`. |
| [handle_write_result](/home/leynos/.lody/repos/github---leynos---cuprum/worktrees/614e3091-8d73-404f-8c2c-a5022f5c0087/rust/cuprum-rust/src/io_utils.rs:33) / [is_nonfatal_write_error](/home/leynos/.lody/repos/github---leynos---cuprum/worktrees/614e3091-8d73-404f-8c2c-a5022f5c0087/rust/cuprum-rust/src/io_utils.rs:21) | proptest | Error taxonomy and saturating total updates are policy-sensitive. | Proves only `BrokenPipe` and `ConnectionReset` are suppressed and success updates totals correctly. | Low. | Return a typed enum: `WriteAction::Continue(bytes)`, `Stop`, `Fail(err)`. |
| [try_splice_pump](/home/leynos/.lody/repos/github---leynos---cuprum/worktrees/614e3091-8d73-404f-8c2c-a5022f5c0087/rust/cuprum-rust/src/splice.rs:43) | Kani, proptest with fake syscalls | Linux splice branch has several rare error policies: `EINVAL`, broken pipe, fatal errors, EOF. | Exhaustively checks dispatch policy and fallback decision boundaries. | Medium-high. Real syscalls are unsuitable for Kani. | Introduce a `SpliceOps` trait or scripted syscall seam. |
| [splice_loop](/home/leynos/.lody/repos/github---leynos---cuprum/worktrees/614e3091-8d73-404f-8c2c-a5022f5c0087/rust/cuprum-rust/src/splice.rs:90) | Kani | Accumulation and termination depend on syscall result sequences. | Proves total equals initial plus successful splice bytes until EOF/error, with drain on non-fatal write error. | Medium after syscall seam. | Extract pure transition over bounded `SpliceResult` traces. |
| [drain_reader](/home/leynos/.lody/repos/github---leynos---cuprum/worktrees/614e3091-8d73-404f-8c2c-a5022f5c0087/rust/cuprum-rust/src/splice.rs:121) | Kani | It currently exits on any `read <= 0`; EINTR/EAGAIN policy is not modelled. | Clarifies whether interrupted/nonblocking reads should retry or terminate. | Medium. | Return `io::Result<()>` and inject read operation for scripted verification. |
| [file_from_raw / pump_stream / consume_stream](/home/leynos/.lody/repos/github---leynos---cuprum/worktrees/614e3091-8d73-404f-8c2c-a5022f5c0087/rust/cuprum-rust/src/lib.rs:130) | Verus only after refactor; Kani for bounded paths | FD ownership invariant is manual: reader is borrowed and forgotten, writer is consumed. | Proves no accidental reader close / double-close in the model and writer closes intentionally. | High. Verus needs an ownership model, not raw `File` effects. | Replace `mem::forget(reader)` with `ManuallyDrop<File>` or typed `BorrowedFd`/`OwnedFd` wrappers. |
| [rust_pump_stream / rust_consume_stream](/home/leynos/.lody/repos/github---leynos---cuprum/worktrees/614e3091-8d73-404f-8c2c-a5022f5c0087/rust/cuprum-rust/src/lib.rs:49) | proptest at Python/Rust boundary | Error mapping and default argument equivalence are currently scenario-tested. | Proves invalid conversion returns `ValueError`, I/O failures map to `OSError`, defaults match explicit values. | Medium. Cross-language tests are slower. | Add internal `RustStreamError` enum and convert to PyO3 errors in one place. |

**Tool Fit**

- **proptest first:** UTF-8 chunking, buffer/FD conversion, consume output equivalence, write-error taxonomy.
- **Kani next:** bounded state machines after seams are added: UTF-8 error transitions, write latch, splice policy, syscall traces.
- **Verus sparingly:** only FD ownership/lifetime modelling is plausibly worth it. Most current code is I/O-bound and not a natural Verus target until pure ownership/state abstractions exist.

**Highest-ROI First Step**

Add Rust unit/property test support and start with `decode_utf8_replace`. It is pure, central to correctness, and likely to find real boundary issues fastest. Then extract event/state seams for `pump_stream_files_readwrite` and `splice_loop`; those are the best Kani candidates.

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

[cuprum/unittests/test_tee_profile_worker.py](https://github.com/leynos/cuprum/pull/62/files/d60e7603cc5b00ab3ea1885e2574f8103e55adf2#diff-7c137723be763ebc802ca9cce96382ea5b54e5046501f5a74686591039259fdb)

Comment on file

~~~python
tests around backend selection, reentrancy guard tests, CLI smoke tests, and
parametrised configuration-validation tests.

The ``_EnvBackendSelector`` concurrency tests mirror the state transitions a
~~~

❌ New issue: [Low Cohesion](https://codescene.io/projects/74471/delta?repo-id=1354509&review-id=62&biomarker=Low+Cohesion&filename=cuprum%2Funittests%2Ftest_tee_profile_worker.py&method=)
This module has at least 15 different responsibilities amongst its 31 functions, threshold = 4

File: cuprum/unittests/test_tee_profile_worker_concurrency.py [hunk 1]
```diff
@@ -184,6 +201,55 @@ def _activate(
...
+
+    @contextlib.contextmanager
+    def _activate(
+        self,
+        backend: tee_profile_worker.BackendName,
+    ) -> cabc.Iterator[None]:
+        """Coordinate lock contention at selector state-machine boundaries."""
+        if backend != "python":
+            self._events["second_waiting_for_lock"].set()
+        with self._delegate(backend):
```
[the remark this answers is not in the document]
Comment (leynos) [reply cuprum/unittests/test_tee_profile_worker_concurrency.py]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai pause

File: cuprum/unittests/test_tee_profile_worker_concurrency.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (leynos) [reply cuprum/unittests/test_tee_profile_worker_concurrency.py]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

[cuprum/unittests/test_tee_profile_worker_concurrency.py](https://github.com/leynos/cuprum/pull/62/files/7495aa3635c006da8eab71a9b83daac64fd18b77#diff-35d29a2828a40aca3318e14fb9659885b2ea5a603e7777adc64cb1123dc5cc3b)

Comment on file

~~~python
"""Tests for ``benchmarks.tee_profile_worker``.
"""Concurrency and reentrancy tests for ``_EnvBackendSelector``.
~~~

❌ New issue: [Global Conditionals](https://codescene.io/projects/74471/delta?repo-id=1354509&review-id=62&biomarker=Global+Conditionals&filename=cuprum%2Funittests%2Ftest_tee_profile_worker_concurrency.py&method=)
There's global code outside of functions with a cyclomatic complexity of 15, threshold = 10

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response. Ensure that this is validated against the *current* version of the codegraph.

If further refinement to address this finding would be deleterious, please supply a clear explanatory one to two paragraph markdown message I can paste into the CodeScene web ui's diagnostic suppression function so this diagnostic can be silenced.

## Global Conditionals
cuprum/unittests/test_tee_profile_worker_concurrency.py: 

### What lead to degradation?
There's global code outside of functions with a cyclomatic complexity of 15, threshold = 10

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?


- In `test_nested_selector_rejects_generated_backend_pairs`, consider reusing `_available_backend_names()` (as you do in `_backend_lists`) for the Hypothesis `outer_backend`/`inner_backend` strategies so the test set respects environments where the Rust backend is not available.
- The hardcoded 5–10 second timeouts on `Event.wait()` and `Thread.join()` in the concurrency tests could lead to occasional flakiness under slow CI; consider centralizing these as constants or slightly relaxing them so they can be tuned in one place if needed.
- The thread orchestration pattern (starting two threads, waiting on events, joining, then asserting on `alive_threads` and collected errors) is duplicated across concurrency tests; extracting a small helper or fixture for this logic would reduce repetition and make the interleaving scenarios easier to extend.

**Comment by leynos** [comment]
@coderabbitai resume

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

### ❌ Failed checks (3 errors, 5 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | Two critical timeout assertions missing: (1) Line 335–337 lacks assertion on event.wait(); (2) Line 108–129 subprocess lacks timeout and capture\_output. | Assert wait() at line 335. Add timeout=30 and capture\_output=True to subprocess.run(), include stderr in assertions. |
| Unit Architecture | ❌ Error | Two architectural defects: unasserted checkpoint timeout at line 335 silently hides lock-order violation; subprocess call lacks timeout/capture\_output, allowing hangs and masking errors. | Assert the wait() result at line 335; add timeout=30, capture\_output=True to subprocess.run(), and assert wait result with stderr in failure message. |
| Security And Privacy | ❌ Error | Two review-flagged issues remain: subprocess.run() lacks timeout/capture\_output (can hang indefinitely); event.wait() lacks result assertion (timeout silently passes). | Add timeout=30, capture\_output=True, text=True to subprocess.run(); assert event.wait() returns True with explicit timeout-failure message. |
| Testing (Unit And Behavioural) | ⚠️ Warning | Test suite is comprehensive, but missing assertions on checkpoint waits (line 335-337) silently bypasses lock-order checks, and subprocess test (line 108) lacks timeout and output capture. | Assert checkpoint wait at line 335; add timeout=30 and capture\_output=True to subprocess.run; include stderr in error message for diagnostics. |
| Testing (Compile-Time / Ui) | ⚠️ Warning | Snapshots are well-designed with focused expectations and explicit nondeterministic field redaction, but unaddressed review comments request critical assertion and timeout handling. | Assert wait() at line 335-337 and add timeout=30, capture\_output=True, text=True to subprocess.run() with stderr in failure message. |
| Performance And Resource Use | ⚠️ Warning | Subprocess test lacks timeout (test\_cli line 108) risking test hangs; unasserted wait() at concurrency line 335 allows silent timeout bypass compromising lock-order test guarantee. | Add timeout=30 and capture\_output=True to subprocess.run(); assert release\_first\_context.wait() result at line 335 with message "expected first context release signal before unlock". |
| Concurrency And State | ⚠️ Warning | Concurrency test Line 335-337: unasserted event.wait() timeout silently passes, defeating lock-order guarantee. CLI test Line 109: missing subprocess timeout risks hanging test run. | Assert `release_first_context.wait()` return value with timeout message at Line 335. Add `timeout=30, capture_output=True, text=True` to subprocess.run at Line 109 and include stderr in assertion. |
| Architectural Complexity And Maintainability | ⚠️ Warning | Two flagged review comments remain unaddressed: missing assert on wait() (Line 335) and missing timeout/capture\_output on subprocess, undermining test reliability. | Address both review comments: assert wait result on Line 335; add timeout=30 and capture\_output=True to subprocess.run with stderr in failure output. |

File: cuprum/unittests/test_tee_profile_worker_concurrency.py [hunk 2]
```diff
@@ -174,16 +283,65 @@ def _activate(
...
+                    timeout=_EVENT_WAIT_TIMEOUT_SECONDS,
+                ), "expected second thread to contend for the backend lock"
+                with self._observation_lock:
+                    self._observations.append(os.environ.get("CUPRUM_STREAM_BACKEND"))
+                assert not self._events["second_entered_context"].is_set(), (
+                    "expected the second thread to block while the first holds the lock"
+                )
+                self._events["release_first_context"].wait(
+                    timeout=_EVENT_WAIT_TIMEOUT_SECONDS,
+                )
```
[the remark this answers is not in the document]
Comment (leynos) [reply cuprum/unittests/test_tee_profile_worker_concurrency.py]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: cuprum/unittests/test_tee_profile_worker_cli.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+            "--repeat-count",
+            "1",
+            "--output",
+            str(output),
+        ],
+        check=False,
+    )
+
+    assert completed.returncode == 0, f"expected exit 0, got {completed.returncode}"
+    assert output.exists(), "expected worker-result.json to be written"
```
[the remark this answers is not in the document]
Comment (leynos) [reply cuprum/unittests/test_tee_profile_worker_cli.py]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

#### cuprum/unittests/test_tee_profile_worker_concurrency.py (1)

`225-232`: _🧹 Nitpick_ | _🔵 Trivial_ | _⚡ Quick win_
**Use the shared thread-join helper in this worker-pair path.**
Replace the local join/alive block at Line 226 with `_join_and_assert_finished` to centralise timeout behaviour and failure text.

##### Proposed patch

```diff
barrier.wait(timeout=_EVENT_WAIT_TIMEOUT_SECONDS)
-    for thread in threads:
-        thread.join(timeout=_THREAD_JOIN_TIMEOUT_SECONDS)
-
-    alive_threads = [thread.name for thread in threads if thread.is_alive()]
-    assert not alive_threads, (
-        f"expected worker threads to finish for {backends}, got {alive_threads}"
-    )
+    _join_and_assert_finished(*threads, context=f"backends={backends!r}")
```

##### 🤖 Detailed instructions

Verify each finding against current code. Fix only still-valid issues, skip the
rest with a brief reason, keep changes minimal, and validate.

In `@cuprum/unittests/test_tee_profile_worker_concurrency.py` around lines 225 -
232, Replace the manual join loop and alive-thread assertion after barrier.wait
with the shared helper _join_and_assert_finished to centralise timeout handling
and consistent failure messages; remove the for thread.join(...) loop and the
alive_threads check and call _join_and_assert_finished(threads, backends) (using
the existing threads list and backends identifier) in their place.

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

### ❌ Failed checks (3 errors, 5 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | Two critical timeout assertions missing: (1) Line 335–337 lacks assertion on event.wait(); (2) Line 108–129 subprocess lacks timeout and capture\_output. | Assert wait() at line 335. Add timeout=30 and capture\_output=True to subprocess.run(), include stderr in assertions. |
| Unit Architecture | ❌ Error | Two architectural defects: unasserted checkpoint timeout at line 335 silently hides lock-order violation; subprocess call lacks timeout/capture\_output, allowing hangs and masking errors. | Assert the wait() result at line 335; add timeout=30, capture\_output=True to subprocess.run(), and assert wait result with stderr in failure message. |
| Security And Privacy | ❌ Error | Two review-flagged issues remain: subprocess.run() lacks timeout/capture\_output (can hang indefinitely); event.wait() lacks result assertion (timeout silently passes). | Add timeout=30, capture\_output=True, text=True to subprocess.run(); assert event.wait() returns True with explicit timeout-failure message. |
| Testing (Unit And Behavioural) | ⚠️ Warning | Test suite is comprehensive, but missing assertions on checkpoint waits (line 335-337) silently bypasses lock-order checks, and subprocess test (line 108) lacks timeout and output capture. | Assert checkpoint wait at line 335; add timeout=30 and capture\_output=True to subprocess.run; include stderr in error message for diagnostics. |
| Testing (Compile-Time / Ui) | ⚠️ Warning | Snapshots are well-designed with focused expectations and explicit nondeterministic field redaction, but unaddressed review comments request critical assertion and timeout handling. | Assert wait() at line 335-337 and add timeout=30, capture\_output=True, text=True to subprocess.run() with stderr in failure message. |
| Performance And Resource Use | ⚠️ Warning | Subprocess test lacks timeout (test\_cli line 108) risking test hangs; unasserted wait() at concurrency line 335 allows silent timeout bypass compromising lock-order test guarantee. | Add timeout=30 and capture\_output=True to subprocess.run(); assert release\_first\_context.wait() result at line 335 with message "expected first context release signal before unlock". |
| Concurrency And State | ⚠️ Warning | Concurrency test Line 335-337: unasserted event.wait() timeout silently passes, defeating lock-order guarantee. CLI test Line 109: missing subprocess timeout risks hanging test run. | Assert `release_first_context.wait()` return value with timeout message at Line 335. Add `timeout=30, capture_output=True, text=True` to subprocess.run at Line 109 and include stderr in assertion. |
| Architectural Complexity And Maintainability | ⚠️ Warning | Two flagged review comments remain unaddressed: missing assert on wait() (Line 335) and missing timeout/capture\_output on subprocess, undermining test reliability. | Address both review comments: assert wait result on Line 335; add timeout=30 and capture\_output=True to subprocess.run with stderr in failure output. |

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai resume

File: docs/developers-guide.md [hunk 4]
```diff
@@ -294,6 +294,89 @@ methods:
...
+
+The worker test suite is split across three focused modules so that each file
+covers one boundary of behaviour:
+
+- `cuprum/unittests/test_tee_profile_worker_core.py` covers parent-side consume
+  hot-path execution, result accounting, and snapshotted worker output.
+- `cuprum/unittests/test_tee_profile_worker_cli.py` covers CLI invocation, the
+  JSON payload shape, and `TeeProfileWorkerConfig` validation errors.
+- `cuprum/unittests/test_tee_profile_worker_concurrency.py` covers reentrancy,
+  lock serialisation, and concurrent-worker race-freedom for
```
[the remark this answers is not in the document]
Comment (leynos) [reply docs/developers-guide.md]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.
